packages/core/src/database/index.ts

```diff
@@ -215,22 +232,58 @@ export const createDatabase = ({
         ? [preBuild.databaseConfig.poolConfig.max - internalMax, 0, 0]
         : [equalMax, equalMax, equalMax];
 
+    const internal = createPool({
+      ...preBuild.databaseConfig.poolConfig,
+      application_name: `${preBuild.namespace}_internal`,
+      max: internalMax,
+      statement_timeout: 10 * 60 * 1000, // 10 minutes to accommodate slow sync store migrations.
+    });
+
+    const connection = (parse as unknown as typeof parse.parse)(
+      preBuild.databaseConfig.poolConfig.connectionString!,
+    );
+
+    const role = `ponder_readonly_${connection.database}_${preBuild.namespace}`;
+
+    await internal.query(`CREATE SCHEMA IF NOT EXISTS "${preBuild.namespace}"`);
+    const hasRole = await internal
+      .query("SELECT FROM pg_roles WHERE rolname = $1", [role])
+      .then(({ rows }) => rows[0]);
+    if (hasRole) {
+      await internal.query(`DROP OWNED BY ${role}`);
+      await internal.query(`DROP ROLE IF EXISTS ${role}`);
+    }
+    await internal.query(`CREATE ROLE ${role} WITH LOGIN PASSWORD 'pw'`);
+    await internal.query(
+      `GRANT CONNECT ON DATABASE "${connection.database}" TO ${role}`,
+    );
+    await internal.query(
+      `GRANT USAGE ON SCHEMA "${preBuild.namespace}" TO ${role}`,
+    );
+    await internal.query(
+      `ALTER DEFAULT PRIVILEGES IN SCHEMA "${preBuild.namespace}" GRANT SELECT ON TABLES TO ${role}`,
+    );
+    await internal.query(
+      `ALTER ROLE ${role} SET search_path TO "${preBuild.namespace}"`,
+    );
+
     driver = {
-      internal: createPool({
-        ...preBuild.databaseConfig.poolConfig,
-        application_name: `${preBuild.namespace}_internal`,
-        max: internalMax,
-        statement_timeout: 10 * 60 * 1000, // 10 minutes to accommodate slow sync store migrations.
-      }),
+      internal,
       user: createPool({
         ...preBuild.databaseConfig.poolConfig,
         application_name: `${preBuild.namespace}_user`,
         max: userMax,
       }),
       readonly: createPool({
         ...preBuild.databaseConfig.poolConfig,
+        connectionString: undefined,
         application_name: `${preBuild.namespace}_readonly`,
         max: readonlyMax,
+        user: role,
+        password: "pw",
+        host: connection.host ?? undefined,
+        port: Number(connection.port!),
+        database: connection.database ?? undefined,
```

Comment:
If this is actually nullable, we might have a problem above in the role name. But I'm pretty sure PG conn strings must explicitly set a database name, so might just be a bad type.